linkerd/proxy/http/src/h2.rs

```diff
@@ -30,17 +33,48 @@ pub struct Connection<B> {
     tx: SendRequest<B>,
 }
 
-pub struct ConnectFuture<F: Future, B> {
+#[pin_project]
+pub struct ConnectFuture<F, B>
+where
+    F: TryFuture,
+    B: HttpBody + Send + 'static,
+    B::Data: Send,
+    B::Error: std::error::Error + Send + Sync,
+    F::Ok: AsyncRead + AsyncWrite + Send + 'static,
+{
+    #[pin]
     state: ConnectState<F, B>,
     h2_settings: Settings,
 }
 
-enum ConnectState<F: Future, B> {
-    Connect(F),
-    Handshake(Handshake<F::Item, B>),
+#[pin_project]
+enum ConnectState<F, B>
+where
+    F: TryFuture,
+    B: HttpBody + Send + 'static,
+    B::Data: Send,
+    B::Error: std::error::Error + Send + Sync,
+    F::Ok: AsyncRead + AsyncWrite + Send + 'static,
+{
+    Connect(#[pin] F),
+    Handshake(
+        #[pin]
+        Pin<
+            Box<
+                dyn Future<
```

Comment:
Is this because `handshake` no longer returns a named type? If so, would it better to add a generic for this TryFuture? I don't have a strong opinion... just trying to figure out what the right idiom is.

Reply:
I realized that if it's unnamed we _have to_ box it. Nevermind

Reply:
yeah, that's correct. eventually, it would be nice to replace this hand-written future with an async block future (on nightly, where we can return `impl Future` from tower services), and thus do away with the box; otherwise, there will always be a box somewhere.